src/JustEat.StatsD/Buffered/StatsDUtf8Formatter.cs

```diff
@@ -24,34 +27,37 @@ public int GetMaxBufferSize(in StatsDMessage msg)
             const int MaxSamplingSuffixSize = 2;
 
             return _utf8Prefix.Length
-                    + Encoding.UTF8.GetByteCount(msg.StatBucket)
-                    + ColonBytes
-                    + MaxSerializedDoubleSymbols
-                    + MaxMessageKindSuffixSize
-                    + MaxSamplingSuffixSize
-                    + MaxSerializedDoubleSymbols;
+                   + Encoding.UTF8.GetByteCount(msg.StatBucket)
+                   + ColonBytes
+                   + MaxSerializedDoubleSymbols
+                   + MaxMessageKindSuffixSize
+                   + MaxSamplingSuffixSize
+                   + MaxSerializedDoubleSymbols
+                   + this._tagsFormatter.GetTagsBufferSize(msg.Tags);
         }
 
         public bool TryFormat(in StatsDMessage msg, double sampleRate, Span<byte> destination, out int written)
         {
             var buffer = new Buffer(destination);
 
             bool isFormattingSuccessful =
-                  TryWriteBucketNameWithColon(ref buffer, msg.StatBucket)
+                  TryWriteBucketNameWithColon(ref buffer, msg)
                && TryWritePayloadWithMessageKindSuffix(ref buffer, msg)
-               && TryWriteSampleRateIfNeeded(ref buffer, sampleRate);
+               && TryWriteSampleRateIfNeeded(ref buffer, sampleRate)
+               && _tagsFormatter.TryWriteSuffixTagsIfNeeded(ref buffer, msg.Tags);
 
             written = isFormattingSuccessful ? buffer.Written : 0;
             return isFormattingSuccessful;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private bool TryWriteBucketNameWithColon(ref Buffer buffer, string statBucket)
+        private bool TryWriteBucketNameWithColon(ref Buffer buffer, in StatsDMessage msg)
```

Comment:
Hi Martin,

I have performed some performance tests with BenchmarkDotNet and the statsd Docker image (<https://hub.docker.com/r/statsd/statsd>).

I downgrade BenchmarkDotNet version to 0.12.0 because they have an issue with the latest lang version project property in the latest version 0.12.1 (https://github.com/dotnet/BenchmarkDotNet/issues/1452).

This is the master branch execution:
![image](https://user-images.githubusercontent.com/8134685/91286686-0a74fc80-e78f-11ea-9a19-10fedaaa9735.png)

With the tagging support, there is some performance loss but the most worrying part is the allocation and performance loss when tags are written:
![image](https://user-images.githubusercontent.com/8134685/91286102-4f4c6380-e78e-11ea-9e3f-dcc0bb344825.png)

Reply:
Sorry, I didn't see this because it was collapsed 😄

Yeah I'm aware of that bug, I fixed it!

Rather than downgrade, can you apply this work around instead?

https://github.com/martincostello/api/blob/f1822671d833e57da83121af17bab92953de6176/tests/API.Benchmarks/API.Benchmarks.csproj#L10-L11

Would you be able to paste the Markdown from the benchmarks into the main body of the PR so they don't get lost?

Then they'll be a bit easier to consume and triage than a screenshot 😃

Reply:
I have applied the workaround and the results are included in the PR body. I didn't know that there was this report in markdown. Good to know!

I didn't see either that you were the fixer :D.